nir/data_ir/graph.py

```diff
@@ -0,0 +1,261 @@
+        """
+
+        if not self.data.dtype == bool:
+            raise ValueError("Data must be boolean to convert to EventData.")
+        idx = np.full((self.n_samples, n_events), -1)
+        time = np.full((self.n_samples, n_events), np.inf)
+
+        if time_shift < 0 or time_shift >= self.dt:
+            raise ValueError("time_shift must be in interval [0, dt)")
+        for sample in range(self.n_samples):
+            time_step, neuron = np.where(self.data[sample])
+
+            order = np.argsort(time_step)  # sort events by time
+            time_step = time_step[order]
+            neuron = neuron[order]
+
+            num_events = min(len(time_step), n_events)
+            idx[sample, :num_events] = neuron[:num_events]
+            time[sample, :num_events] = time_step[:num_events] * self.dt + time_shift
+
+        return EventData(idx, time, self.n_neurons, self.t_max)
+
+
+@dataclass
+class EventData:
+    """
+    Event-based data represented as a list of event indices and their
+    corresponding timestamps. Each event is discrete and carries no magnitude;
+    it is defined solely by its occurrence at a certain time.
+
+    Arguments
+    ---------
+    idx : np.ndarray[int], shape (n_samples, n_events)
+        Event indices. If there is no event, the index is `-1`.
+    time : np.ndarray[float], shape (n_samples, n_events)
+        Event times. If there is no event, the time is `np.inf`.
+    n_neurons : int
+        Total number of neurons in the layer.
+    t_max : float
+        Maximum time of the recording.
+    """
+
+    idx: np.ndarray
+    time: np.ndarray
+    n_neurons: int
+    t_max: float
+
+    def __post_init__(self):
+        if self.idx.shape != self.time.shape:
+            raise ValueError("idx and time must have the same shape")
+
+    @property
+    def shape(self):
+        return self.idx.shape
+
+    @property
+    def n_samples(self):
+        return self.idx.shape[0]
+
+    def to_time_gridded(
+        self, dt: float  # pylint: disable=invalid-name
+    ) -> TimeGriddedData:
+        """
+        Arguments
+        ---------
+        dt : float
+            Time step size.
+        """
+        n_time_steps = int(self.t_max / dt)
+        discrete_data = np.zeros(
+            (self.n_samples, n_time_steps, self.n_neurons), dtype=bool
+        )
+
+        for sample in range(self.n_samples):
+            valid_spikes = self.idx[sample] != -1
+            valid_times = self.time[sample][valid_spikes]
+            steps = np.floor((valid_times / dt)).astype(int)
+            neurons = self.idx[sample][valid_spikes]
+            discrete_data[sample, steps, neurons] = True
+        return TimeGriddedData(discrete_data, dt)
+
+
+@dataclass
+class ValuedEventData(EventData):
+    """
+    Valued event-based data as a list of event indices, event times and event
+    values.
+
+    Parameters
+    ----------
+    idx : np.ndarray[int], shape (n_samples, n_events)
```

Comment:
These parameters don't seem to align with the constructor of the dataclass (`values`). In fact, where are the `idx` and `time` class variables?! Is the `test_generate_valued_event_data` test passing?